### PR TITLE
Preserve last used units in the Journal activity

### LINKED ISSUE #13: Save last used dimension and units in journal object
Reproducer: start, change dimension and units, stop, restart from journal entry.

Expected: dimension and units preserved.

Observed: dimension and units are reset to Length - Cables.

At the same time, add a reset button to restore dimension and units to defaults.

### PR DESCRIPTION
**Pull request by nswarup14**
Fixes #13 

Kindly review

### CODE REVIEW

File: activity.py [hunk 1]
```diff
@@ -224,7 +242,11 @@ def __init__(self, handle):
         toolbarbox.toolbar.insert(stopbtn, -1)
 
         self.set_toolbar_box(toolbarbox)
-        self._update_combo(convert.length)
+        if 'state' in self.metadata:
```
Comment (chimosky) [inline activity.py]: L245-L249 should be in `read_file`.

**Review by chimosky — CHANGES_REQUESTED** [review]

**Comment by chimosky** [comment]
Tested, comments.

The units aren't preserved, the `combobox` items show the last units accessed but both units show the first item in the `combobox` and the convert values aren't shown - the value being converted and the output of the conversion -.

File: activity.py [hunk 2]
```diff
@@ -251,13 +273,23 @@ def _update_label(self, entry):
         except ValueError:
             pass
 
-    def _update_combo(self, data):
+    def _update_combo(self, unit_type):
+        self._active_unit = unit_type
```
Comment (chimosky) [inline activity.py]: Is there a use for this?.

File: activity.py [hunk 3]
```diff
@@ -251,13 +273,23 @@ def _update_label(self, entry):
...
-        self.dic = data
+        self.dic = units[unit_type]
         keys = self.dic.keys()
         keys.sort()
+
         for x in keys:
             self._liststore.append(['%s' % (x)])
+
+        if self.state is not None:
+            print('i entered')
```
Comment (chimosky) [inline activity.py]: Could this be more informative, probably use `logging` instead of `print`.

File: activity.py (hunk 2 above)
Comment (nswarup14) [reply activity.py]: @chimosky
This is to keep track of what type of units are currently being used.
If the user changes from `length` to `area`, the value is updated. 
When the user closes the activity, this information is stored in the journal object.

File: activity.py (hunk 3 above)
Comment (nswarup14) [reply activity.py]: Sorry
That was actually a debug message. I shall remove it.

**Comment by quozl** [comment]
Tested f988ce5.

* [ ] the unit type (dimension) is not restored on resume,
* [ ] the unit is saved as a number not a name, which means any changes to units in the program may cause journal objects to silently corrupt when the activity is upgraded.

The first point is an obvious functional failure which you should have seen in testing.  Please test before commit, or if the work is not finished mark the pull request as draft, thanks.

**Comment by quozl** [comment]
I'm working on rebasing this.  Should finish tomorrow.

**Comment by quozl** [comment]
@chimosky, @nswarup14, please test 5c1dbd3.  Yet to finish rebase, but it is in a working state for me.

**Comment by nswarup14** [comment]
@quozl @chimosky 
Tested https://github.com/sugarlabs/convert/commit/5c1dbd3f99190ceaeea92df0c58bc2f0e0e6e974 on Sugar v0.114, Ubuntu 18.04.

Individual units for both the combo boxes are being preserved.  The dimension is being preserved as well.
However, the RadioToolButton at the toolbar is not updated, on activity resume.

For example, if I save the activity with `Area` as dimension, on activity resume, the `length` RadioToolButton is set active.

I will try to fix this.

Thanks!

**Comment by nswarup14** [comment]
@quozl @chimosky 
As mentioned in the earlier comment, I have fixed the problem of RadioToolButton not updating with the preserved dimension.

Kindly review.

Thanks

**Comment by quozl** [comment]
Thanks.  Reworked.  Please test 4c232d5.  I'll plan to rebase tomorrow my time.

**Comment by nswarup14** [comment]
@quozl 
Tested https://github.com/sugarlabs/convert/commit/4c232d5c0b67a136dc81491db1ddb8e39e063a23
on Sugar v0.114
Works as expected. Logs contain no errors.

**Comment by quozl** [comment]
Rebased as 5ad6553.

@Aniket21mathur, I've checked and none of the code you wrote in https://github.com/sugarlabs/convert/pull/23/files seems to have been used.  Let me know if you think it has, and I'll adjust the _Co-authored-by_ tag.

I thought it was all finished, but no.  Yet to be done;

* [ ] a button to reset dimension and units to default, per https://github.com/sugarlabs/convert/issues/13, unless we can agree it isn't needed,
* [ ] discuss whether it makes sense to save the _from value_ and _to value_ in the journal object.

**Comment by nswarup14** [comment]
@quozl @chimosky 

> a button to reset dimension and units to default, per #13, unless we can agree it isn't needed,

I feel we can add a button which says reset to SI unit for each corresponding dimension.

> discuss whether it makes sense to save the from value and to value in the journal object.

Yes it does. If we are saving the last used units and dimensions, the next step would be to save the last converted values. If a user wishes to pick up from where he left, then the feature would be ideal.

**Comment by chimosky** [comment]
Tested 5ad6553.

Works as expected. 

> - [ ] a button to reset dimension and units to default, per #13, unless we can agree it isn't needed,

I think we can add such button, seems like a cool feature.
> - [ ] discuss whether it makes sense to save the from value and to value in the journal object.

I think it makes sense to save the from value and to value in the journal object.

**Comment by quozl** [comment]
>  a button to reset dimension and units to default, per #13, unless we can agree it isn't needed,

My vote is no.  As of 5ad65539e4eb6f2391d72481d1850bbc0c1464ec each time a dimension radio tool button is pressed, the unit is already reset to default for that dimension.  Every button is a reset.  Also, per https://github.com/sugarlabs/convert/issues/33 there is no room left on the toolbar.  What do you think?

> discuss whether it makes sense to save the _from value_ and _to value_ in the journal object.

My vote is yes for this one.  In ad63d02.  Please test.

Fixes #24.

**Comment by nswarup14** [comment]
@quozl @chimosky 
Tested and reviewed https://github.com/sugarlabs/convert/commit/ad63d02e5f8f39b3eab86d7e83ff3d56f8a2a26d.
The activity stores the _from_value_ and _to_value_ correctly.

> My vote is no. 

Based on the reasoning provided, I agree with @quozl . The button seems unnecessary as every dimension button serves as a reset button.

**Pull request merged**